core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
I just looked through all the `put*` methods in `BlockManager`. It seems that we always call these methods with `tellMaster = true` for non-broadcast blocks. This would suggest that the driver should have the most updated states for the blocks that we care about here.

I think @lianhuiwang is correct. An executor RPC does not seem necessary here, but I would test it just to be sure (persist, unpersist, evict old blocks...).